sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,621 @@
+
+                centroid_window_first_loading_feature_idx = 0
+
+                for _2 in range(centroids_window_height_ratio_multiplier):
+                    window_loading_feature_idx = (
+                        centroid_window_first_loading_feature_idx
+                        + window_loading_feature_offset
+                    )
+                    loading_feature_idx = first_feature_idx + window_loading_feature_idx
+
+                    if (loading_feature_idx < n_features) and (
+                        loading_centroid_idx < n_clusters
+                    ):
+                        value = current_centroids_t[
+                            loading_feature_idx, loading_centroid_idx
+                        ]
+                    else:
+                        value = zero
+
+                    centroids_window[
+                        window_loading_feature_idx, window_loading_centroid_idx
+                    ] = value
+
+                    centroid_window_first_loading_feature_idx += (
+                        nb_of_window_features_per_work_group
+                    )
+
+                dpex.barrier()
+
+                for window_feature_idx in range(window_nb_features):
+                    feature_idx = window_feature_idx + first_feature_idx
+                    if (feature_idx < n_features) and (sample_idx < n_samples):
+                        # performance for the line thereafter relies on L1 cache
+                        X_value = X_t[feature_idx, sample_idx]
+                    else:
+                        X_value = zero
+                    for window_centroid_idx in range(window_nb_centroids):
+                        centroid_value = centroids_window[
+                            window_feature_idx, window_centroid_idx
+                        ]
+                        partial_scores[window_centroid_idx] += centroid_value * X_value
+
+                dpex.barrier()
+
+                first_feature_idx += window_nb_features
+
+            for i in range(window_nb_centroids):
+                current_score = centroids_window_half_l2_norm[i] - partial_scores[i]
+                if current_score < min_score:
+                    min_score = current_score
+                    min_idx = first_centroid_idx + i
+
+            dpex.barrier()
+
+            first_centroid_idx += window_nb_centroids
+
+        if sample_idx >= n_samples:
+            return
+
+        pseudo_inertia[sample_idx] = min_score
+
+        privatization_idx = (
+            sample_idx // preferred_work_group_size_multiple
+        ) % nb_centroids_private_copies
+        dpex.atomic.add(
+            centroid_counts_private_copies,
+            (privatization_idx, min_idx),
+            one,
+        )
+
+        for feature_idx in range(n_features):
+            dpex.atomic.add(
+                centroids_t_private_copies,
+                # ((privatization_idx + feature_idx) % nb_centroids_private_copies, feature_idx, min_idx),
```

Comment:
I do not understand this comment: is this comment left here as an alternative?

Reply:
i'll remove it, it was here for exploration of the privatization strategy but I'm not sure it makes any sense to try